agent/api/task/task.go

```diff
@@ -2338,64 +2338,135 @@ func (task *Task) dockerLinks(container *apicontainer.Container, dockerContainer
+
+	// bridge network mode and service connect enabled task
 	if task.IsServiceConnectEnabled() && task.IsNetworkModeBridge() {
 		if container.Type == apicontainer.ContainerCNIPause {
-			// we will create bindings for task containers (including both customer containers and SC Appnet container)
-			// and let them be published by the associated pause container.
-			// Note - for SC bridge mode we do not allow customer to specify a host port for their containers. Additionally,
-			// When an ephemeral host port is assigned, Appnet will NOT proxy traffic to that port
+			// we will create bindings for task containers (including both customer containers and service connect AppNet container)
+			// and let them be published by the associated pause containers.
+			// Note - for service connect bridge mode, we do not allow customer to specify a host port for their containers.
+			// Additionally, AppNet will NOT proxy traffic to that port when an ephemeral host port is assigned.
 			taskContainer, err := task.getBridgeModeTaskContainerForPauseContainer(container)
 			if err != nil {
 				return nil, err
 			}
+
+			scContainer := task.GetServiceConnectContainer()
 			if taskContainer == scContainer {
-				// create bindings for all ingress listener ports
-				// no need to create binding for egress listener port as it won't be access from host level or from outside
-				for _, ic := range task.ServiceConnectConfig.IngressConfig {
-					listenerPortInt := int(ic.ListenerPort)
-					dockerPort := nat.Port(strconv.Itoa(listenerPortInt)) + "/tcp"
-					hostPort := 0           // default bridge-mode SC experience - host port will be an ephemeral port assigned by docker
-					if ic.HostPort != nil { // non-default bridge-mode SC experience - host port specified by customer
-						hostPort = int(*ic.HostPort)
-					}
-					dockerPortMap[dockerPort] = append(dockerPortMap[dockerPort], nat.PortBinding{HostPort: strconv.Itoa(hostPort)})
-					// append non-range, singular container port to the containerPortSet
-					containerPortSet[listenerPortInt] = struct{}{}
-					// set taskContainer.ContainerPortSet to be used during network binding creation
-					taskContainer.SetContainerPortSet(containerPortSet)
+				// for service connect container, we need to create port bings and construct dockerPortMap
+				// and containerPortSet with it's ingress config. No need to do this for egress listener port
+				// as it won't be access from host level or from outside
+				dockerPortMap, containerPortSet, err := task.updatePortMapWithSCIngressConfig(dynamicHostPortRange)
+				if err != nil {
+					return nil, err
 				}
+				// set taskContainer.ContainerPortSet to be used during network binding creation
+				taskContainer.SetContainerPortSet(containerPortSet)
 				return dockerPortMap, nil
+			} else {
+				// for non-service connect container, dockerPortMap will be updated in the later
+				containerToCheck = taskContainer
```

Comment:
Q: Wondering if a task contains multiple task container is a valid case and how are we handling it?